### PR TITLE
An initial stab at Airlift documentation

### PR DESCRIPTION
**Pull request by Randgalt**
I used Dropwizard as a template. Note: this is only a first attempt.
If you like this direction I will continue otherwise let me know what
to do differently.

Click here to see it rendered: https://github.com/Randgalt/airlift/blob/jordanz/intial-take-on-docs/README.md

cc @martint

### CODE REVIEW

File: docs/ref_lifecycle.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[◀︎ Airlift](../README.md) • [◀︎ Reference](../README.md#reference)
+
+## Reference: Lifecycle/Bootstrapping
+
+### @PostConstruct and @PreDestroy
+
+Airlift enables the standard [@PostConstruct](https://docs.oracle.com/javase/8/docs/api/javax/annotation/PostConstruct.html) and [@PreDestroy](https://docs.oracle.com/javase/8/docs/api/javax/annotation/PreDestroy.html) annotations for objects injected/bound
```
Comment (electrum) [inline docs/ref_lifecycle.md:7]: Do we want to use code quotes inside links
```markdown
[`@PostConstruct`](https://docs.oracle.com/javase/8/docs/api/javax/annotation/PostConstruct.html)

File: docs/recipes.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+{
+    protected void setup(Binder binder)
+    {
+        // NOTE: this method also binds the MyConfig
+        install(installModuleIf(MyConfig.class, MyConfig::getSomeValue, new MyConditionalModule()));
+    }
+}
+```
+
+### How Do I Map Environment Values to Config Objects
```
Comment (electrum) [inline docs/recipes.md:37]: Probably remove, since the environment support is for values, not config keys

**Review by electrum — COMMENTED** [review]
Looks great so far. Some random comments

File: docs/ref_lifecycle.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+[◀︎ Airlift](../README.md) • [◀︎ Reference](../README.md#reference)
+
+## Reference: Lifecycle/Bootstrapping
+
+### @PostConstruct and @PreDestroy
```
[the remark this answers is not in the document]
Comment (Randgalt) [reply docs/ref_lifecycle.md:5]: Yeah - I tested it

File: docs/ref_lifecycle.md (hunk 1 above)
Comment (Randgalt) [reply docs/ref_lifecycle.md:7]: I tried that and the formatting was odd

**Comment by Randgalt** [comment]
@electrum I think I addressed all your comments.

File: README.md [hunk 4]
```diff
@@ -5,3 +5,37 @@
 Airlift is a framework for building REST services in Java.
 
 This project is used as the foundation for distributed systems like [Trino (formerly PrestoSQL)](https://trino.io).
+
+Airlift pulls together stable, mature libraries from the Java ecosystem into a simple, light-weight package that lets you focus on getting things done and includes built-in support for configuration, metrics, logging, IoC, and much more, enabling you and your team to ship a production-quality web service in the shortest time possible.
```
Comment (martint) [inline README.md:9]: IoC -> dependency injection

File: docs/add_logging.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+### Step 2 - Start Logging
+
+Example logging:
+
+```java
+import io.airlift.log.Logger;
+
+public class MyClass
+{
+    private static final Logger log = Logger.get(MyClass.class);
```
Comment (martint) [inline docs/add_logging.md:33]: LOG

File: docs/add_metrics.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+<dependency>
+    <groupId>io.airlift</groupId>
+    <artifactId>jmx-http-rpc</artifactId>
+</dependency>
+
+<dependency>
+    <groupId>io.airlift</groupId>
+    <artifactId>jmx-http</artifactId>
+</dependency>
```
Comment (martint) [inline docs/add_metrics.md:32]: These are only needed if you want to expose JMX metrics via HTTP, so I'd call that out or simply remove it or move it to a different tutorial.

File: docs/overview.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
+[◀︎ Airlift](../README.md)
+
+## Overview
+
+Airlift is a library and not a framework. Frameworks such as Spring are very heavy-weight and will control the future of your project. Libraries, on the other hand, are small tools that enable your project by providing targeted utility. Think of Airlift as plumbing for your service. Airlift helps you by eliminating the drudgery of piecing together the basics while allowing you to focus on the functionality of your service.
```
Comment (martint) [inline docs/overview.md:5]: I like to think of Airlift as a "toolkit" -- a set of libraries that work well together (or are made to work well together via adapters and integrations)

**Review by martint — COMMENTED** [review]
This is good stuff! I just had a few minor comments.

**Comment by Randgalt** [comment]
@martint your comments have been addressed

File: docs/overview.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
+[◀︎ Airlift](../README.md)
+
+## Overview
+
+Airlift is a toolkit and not a framework. Frameworks such as Spring are very heavy-weight and will control the future of your project. Libraries, on the other hand, are small tools that enable your project by providing targeted utility. Think of Airlift as plumbing for your service. Airlift helps you by eliminating the drudgery of piecing together the basics while allowing you to focus on the functionality of your service.
```
Comment (martint) [inline docs/overview.md:5]: The third sentence still refers to libraries, so it sounds disconnected from the introduction.

Comment (Randgalt) [reply docs/overview.md]: done

**Pull request merged**